pumpkin-world/src/world_gen/noise/perlin.rs

```diff
@@ -0,0 +1,1134 @@
+use itertools::Itertools;
+use num_traits::{Pow, Zero};
+use pumpkin_core::random::{RandomDeriverImpl, RandomGenerator, RandomImpl};
+
+use super::{dot, lerp3, GRADIENTS};
+
+pub struct PerlinNoiseSampler {
+    permutation: Box<[u8]>,
+    x_origin: f64,
+    y_origin: f64,
+    z_origin: f64,
+}
+
+impl PerlinNoiseSampler {
+    pub fn new(random: &mut impl RandomImpl) -> Self {
+        let x_origin = random.next_f64() * 256f64;
+        let y_origin = random.next_f64() * 256f64;
+        let z_origin = random.next_f64() * 256f64;
+
+        let mut permutation = [0u8; 256];
+
+        permutation
+            .iter_mut()
+            .enumerate()
+            .for_each(|(i, x)| *x = i as u8);
+
+        for i in 0..256 {
+            let j = random.next_bounded_i32((256 - i) as i32) as usize;
+            permutation.swap(i, i + j);
+        }
+
+        Self {
+            permutation: Box::new(permutation),
+            x_origin,
+            y_origin,
+            z_origin,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn sample_flat_y(&self, x: f64, y: f64, z: f64) -> f64 {
+        self.sample_no_fade(x, y, z, 0f64, 0f64)
+    }
+
+    pub fn sample_no_fade(&self, x: f64, y: f64, z: f64, y_scale: f64, y_max: f64) -> f64 {
+        let trans_x = x + self.x_origin;
+        let trans_y = y + self.y_origin;
+        let trans_z = z + self.z_origin;
+
+        let x_int = trans_x.floor() as i32;
+        let y_int = trans_y.floor() as i32;
+        let z_int = trans_z.floor() as i32;
+
+        let x_dec = trans_x - x_int as f64;
+        let y_dec = trans_y - y_int as f64;
+        let z_dec = trans_z - z_int as f64;
+
+        let y_noise = if y_scale != 0f64 {
+            let raw_y_dec = if y_max >= 0f64 && y_max < y_dec {
+                y_max
+            } else {
+                y_dec
+            };
+            (raw_y_dec / y_scale + 1.0E-7f32 as f64).floor() * y_scale
+        } else {
+            0f64
+        };
```

Comment:
This part be made branchless:
```suggestion
let c = y_max >= 0f64 && y_max < y_dec;
let y_noise = ( ((y_scale as u64 & 1f64 as u64) as f64) * (((check as u8 as u64 as f64) * y_max) + (!check as u8 as u64 as f64 * y_dec)) / y_scale + 1.0E-7f32 as f64).floor() * y_scale;
```

I've made quick benchmark with criterion using completely randomly generated float pointer number and the there is 17% improvements. I'm not sure however what data are input into this. @kralverde If you believe that each branch have the same odd of being taken, then the brancheless version is better. Otherwise stay with the branch version.

```
branch                  time:   [18.688 ns 18.706 ns 18.724 ns]
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low mild
  7 (7.00%) high mild

brancheless             time:   [15.963 ns 15.976 ns 15.989 ns]
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe
```

Reply:
!=0 should be more likely, ill make it branchless

Reply:
The recommended code gives incorrect results

Reply:
@kralverde I have just tested too. I apologize for proposing a wrong solution.

I've come up with a fixed version but that is performing worse than the branch one:
```rust
let c = y_max >= 0f64 && y_max < y_dec;
    (((1u64 ^ y_scale as u64 + !(y_scale as u64) & !(y_scale as u64)) >> 63) as f64) * ( ( (((c as u8 as u64 as f64) * y_max) + (!c as u8 as u64 as f64 * y_dec) ) / y_scale + 1.0E-7f32 as f64).floor() * y_scale )
```
```
branch                  time:   [21.595 ns 21.602 ns 21.610 ns]
brancheless             time:   [24.572 ns 24.582 ns 24.592 ns]
```

You can forget it. Thank you again for noticing.